pytest/tests/mocknet/locust.py

```diff
@@ -0,0 +1,168 @@
+
+
+def upload_key(node, filename):
+    node.machine.upload(args.funding_key,
+                        '/home/ubuntu/locust/funding_key.json',
+                        switch_user='ubuntu')
+
+
+def run_master(args, node):
+    upload_key(node, args.funding_key)
+    cmd = f'/home/ubuntu/locust/venv/bin/python3 -m locust --web-port 3030 --master-bind-port 3000 -H {args.node_ip_port} -f locustfiles/{args.locustfile} --shard-layout-chain-id mainnet --funding-key=/home/ubuntu/locust/funding_key.json --max-workers {args.max_workers} --master'
+    cmd_utils.run_in_background(
+        node,
+        cmd,
+        'locust-master.txt',
+        pre_cmd=
+        'ulimit -S -n 100000 && cd /home/ubuntu/nearcore/pytest/tests/loadtest/locust'
+    )
+
+
+def wait_locust_inited(node, log_filename):
+    # We want to wait for the locust process to finish the initialization steps. Is there a better way than
+    # just waiting for the string "Starting Locust" to appear in the logs?
+    cmd_utils.run_cmd(
+        node,
+        f'tail -f {cmd_utils.LOG_DIR}/{log_filename}.0 | grep --line-buffered -m 1 -q "Starting Locust"'
+    )
+
+
+def wait_master_inited(node):
+    wait_locust_inited(node, 'locust-master.txt')
+    logger.info(f'master locust node initialized')
+
+
+def wait_worker_inited(node):
+    wait_locust_inited(node, 'locust-worker.txt')
+    logger.info(f'worker locust node {node.instance_name} initialized')
+
+
+def run_worker(args, node, master_ip):
+    cmd = f'/home/ubuntu/locust/venv/bin/python3 -m locust --web-port 3030 -H {args.node_ip_port} -f locustfiles/{args.locustfile} --shard-layout-chain-id mainnet --funding-key=/home/ubuntu/locust/funding_key.json --worker --master-port 3000'
+    if master_ip != node.machine.ip:
+        # if this node is also the master node, the key has already been uploaded
+        upload_key(node, args.funding_key)
+        cmd += f'--master-host {master_ip}'
+    cmd_utils.run_in_background(
+        node,
+        cmd,
+        'locust-worker.txt',
+        pre_cmd=
+        'ulimit -S -n 100000 && cd /home/ubuntu/nearcore/pytest/tests/loadtest/locust'
+    )
+
+
+def run_cmd(args):
+    master, workers = parse_instsance_names(args)
+
+    run_master(args, master)
+    wait_master_inited(master)
+    pmap(lambda n: run_worker(args, n, master.machine.ip), workers)
+    pmap(wait_worker_inited, workers)
+    logger.info(
+        f'all locust workers initialized. visit http://{master.machine.ip}:3030/ to start and control the test'
+    )
+
+
+def stop_cmd(args):
+    master, workers = parse_instsance_names(args)
+    # TODO: this feels kind of imprecise and heavy-handed, since we're just looking for a command that matches "python3.*locust.*master" and killing it,
+    # instead of remembering what the process' IP was. Should be possible to do this right, but this will work for now
+    cmd_utils.run_cmd(
+        master,
+        'kill $(ps -C python3 -o pid=,cmd= | grep "locust" | cut -d " " -f 2)')
```

Comment:
looks like `killall` could also be an option to consider: https://man7.org/linux/man-pages/man1/killall.1.html

Reply:
so I took a look at it, but looks like I can't filter by the python programs that have "locust" in their command lines right? Unless I'm missing an option to do that. So that way, it would sadly be even more heavy-handed than what we have here, since we'd just be killing all python3 programs on the machine. I think in the future it would be good to just do this properly and somehow remember what process we started

Reply:
Doesn't `killall -r locust` work?

Reply:
they're all started with `python3 -m locust`, so they don't have `locust` as `comm` and it won't find them with that command sadly